pkg/testers/ginkgo/package.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	return t.extractBinaries(downloadPath)
+}
+
+func (t *Tester) extractBinaries(downloadPath string) error {
+	// finally, search for the test package and extract it
+	f, err := os.Open(downloadPath)
+	if err != nil {
+		return fmt.Errorf("failed to open downloaded tar at %s: %s", downloadPath, err)
+	}
+	defer f.Close()
+	gzf, err := gzip.NewReader(f)
+	if err != nil {
+		return fmt.Errorf("failed to create gzip reader: %s", err)
+	}
+	tarReader := tar.NewReader(gzf)
+	// this is the expected path of the package inside the tar
+	// it will be extracted to e2eTestPath in the loop
+	testPackagePath := "kubernetes/test/bin/e2e.test"
+	foundTestPackage := false
+	// likewise for the actual ginkgo binary
+	binaryPath := "kubernetes/test/bin/ginkgo"
+	foundBinary := false
+	for {
+		// Put this check before any break condition so we don't
+		// accidentally incorrectly error
+		if foundTestPackage && foundBinary {
+			return nil
+		}
+
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("error during tar read: %s", err)
+		}
+
+		if header.Name == testPackagePath {
+			outFile, err := os.Create(e2eTestPath)
+			if err != nil {
+				return fmt.Errorf("error creating file at %s: %s", e2eTestPath, err)
+			}
+			defer outFile.Close()
+
+			if err := outFile.Chmod(0700); err != nil {
+				return fmt.Errorf("failed to make %s executable: %s", e2eTestPath, err)
+			}
+
+			if _, err := io.Copy(outFile, tarReader); err != nil {
+				return fmt.Errorf("error reading data from tar with header name %s: %s", header.Name, err)
+			}
+
+			foundTestPackage = true
+		} else if header.Name == binaryPath {
+			outFile, err := os.Create(binary)
+			if err != nil {
+				return fmt.Errorf("error creating file at %s: %s", binary, err)
+			}
+			defer outFile.Close()
+
+			if err := outFile.Chmod(0700); err != nil {
+				return fmt.Errorf("failed to make %s executable: %s", binary, err)
+			}
+
+			if _, err := io.Copy(outFile, tarReader); err != nil {
+				return fmt.Errorf("error reading data from tar with header name %s: %s", header.Name, err)
+			}
+
+			foundBinary = true
+		}
+	}
+	if !foundBinary && !foundTestPackage {
+		return fmt.Errorf("failed to find %s or %s in %s", binaryPath, testPackagePath, downloadPath)
+	}
+	if !foundBinary {
+		return fmt.Errorf("failed to find %s in %s", binaryPath, downloadPath)
+	}
+	return fmt.Errorf("failed to find %s in %s", testPackagePath, downloadPath)
+}
+
+// ensureReleaseTar checks if the kubernetes test tarball already exists
+// and verifies the hashes
+// else downloads it from GCS
+func (t *Tester) ensureReleaseTar(downloadPath, releaseTar string) error {
+	if _, err := os.Stat(downloadPath); err == nil {
+		klog.V(0).Infof("Found existing tar at %v", downloadPath)
+		if err := t.compareSHA(downloadPath, releaseTar); err == nil {
+			klog.V(0).Infof("Validated hash for existing tar at %v", downloadPath)
+			return nil
+		}
+		klog.Warning(err)
```

Comment:
I think it would be better to explicitly check if the err with `os.IsNotExist(err)` and log a non-warning message if the problem is simply that it needs to be downloaded. If it is a different error, then this should fail immediately.

Reply:
I think in either case we want to proceed with downloading? which will overwrite if hashes don't match etc.

Reply:
True. Just a little worried about a different error suggesting some filesystem issue, but I imagine we would have caught it already? Should be fine.